runtime/src/main/scala/akka/grpc/internal/ClientState.scala

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.grpc.internal
+
+import java.util.concurrent.atomic.AtomicReference
+
+import akka.Done
+import akka.annotation.InternalApi
+import akka.grpc.GrpcClientSettings
+import akka.stream.{ ActorMaterializer, Materializer }
+import io.grpc.ManagedChannel
+
+import scala.annotation.tailrec
+import scala.concurrent.{ ExecutionContext, Future, Promise }
+import scala.util.Failure
+
+/**
+ * INTERNAL API
+ * Used from generated code so can't be private.
+ *
+ * Client utilities taking care of Channel reconnection and Channel lifecycle in general.
+ */
+@InternalApi
+class ClientState(settings: GrpcClientSettings)(implicit mat: Materializer, ex: ExecutionContext) {
+
+  private val internalChannelRef: AtomicReference[InternalChannel] = new AtomicReference[InternalChannel](create())
+  private val closeDemand: Promise[Done] = Promise[Done]()
+
+  mat match {
+    case m: ActorMaterializer =>
+      m.system.whenTerminated.foreach(_ => close())(ex)
+    case _ =>
+  }
+
+  def withChannel[A](f: Future[ManagedChannel] => A): A = {
+    val c = internalChannelRef.get()
+    if (c != null)
+      f(internalChannelRef.get().managedChannel)
+    else // Likely to happen if this is a shared client during shutdown
+      throw new ClientClosedException
+  }
+
+  def closed(): Future[Done] = {
+    // while there's no request to close this RestartingClient, it will continue to restart.
+    // Once there's demand, the `closed` future will redeem flatMapping with the `closed()`
+    // future of the clientRef that's active at that moment.
+    closeDemand.future.flatMap { _ =>
+      val c = internalChannelRef.get()
+      if (c != null)
+        internalChannelRef.get().done
+      else
+        Future.successful(Done)
+    }
+  }
+
+  @tailrec
+  final def close(): Future[Done] = {
+    closeDemand.trySuccess(Done)
+    val c = internalChannelRef.get()
+    if (c != null) {
+      val done = ChannelUtils.close(c)
```

Comment:
Two concurrent calls could both get here and cause 2 calls to close, is that ok? If not we likely will need to have another closing state cas and only let the winner close it.

Reply:
Yeah, I think that would fix it, only one can win the cas for null.

Reply:
> f not we likely will need to have another closing state cas and only let the winner close it.

Hmm, that sounds right.

I think the intermediate state is actually the future returned by `ChannelUtils.close(c)` so that when `AtomicRef[OptionVal[Future[Done]]]` is defined, the value is `Future[Done]` completes when the channel actually closes.

Reply:
close calls shutdown on the ManagedChannel which is idempotent

Reply:
True, but I think there's a problem with what happens after that: 

```scala
final def close(): Future[Done] = {
    closeDemand.trySuccess(Done)
    val c = internalChannelRef.get()
    if (c != null) {
      val done = ChannelUtils.close(c)
      if (internalChannelRef.compareAndSet(c, null))
        done
      else
        close()
    } else
      Future.successful(Done)
  }
```

Two concurrent calls will compete for nullifying `internalChannelRef`, one will win and get `done` and the other will recurse. On the recursion, it'll get `Future.successful(Done)` which is wrong since the closing operation hasn't finished yet.

If we add that intermediate cas, on the recursion it could get `done` too.